engine/src/deposit_event.rs

```diff
@@ -1,11 +1,105 @@
+    Eth {
+        receiver_id: AccountId,
+        message: String,
+    },
+}
+
+impl TokenMessageData {
+    /// Parse event message data for tokens. Data parsed form event `recipient` field.
+    /// Used for Deposit flow.
+    /// For Eth logic flow message validated and prepared for  `ft_on_transfer` logic.
+    /// It mean validating Eth address correctness and preparing message for
+    /// parsing for `ft_on_transfer` message parsing with correct and validated data.
+    pub fn parse_event_message_and_prepare_token_message_data(
+        message: &str,
+        fee: Fee,
+    ) -> Result<TokenMessageData, error::ParseEventMessageError> {
+        let data: Vec<_> = message.split(':').collect();
+        // Data array can contain 1 or 2 elements
+        if data.len() >= 3 {
+            return Err(error::ParseEventMessageError::TooManyParts);
+        }
+        let account_id = AccountId::try_from(data[0].as_bytes())
+            .map_err(|_| error::ParseEventMessageError::InvalidAccount)?;
+
+        // If data array contain only one element it should return NEAR account id
+        if data.len() == 1 {
+            Ok(TokenMessageData::Near(account_id))
+        } else {
+            let raw_message = data[1].into();
+            let message = Self::prepare_message_for_on_transfer(&account_id, fee, raw_message)?;
+
+            Ok(TokenMessageData::Eth {
+                receiver_id: account_id,
+                message,
+            })
+        }
+    }
+
+    // Get recipient account id from Eth part of Token message data
+    pub fn get_recipient(&self) -> String {
+        match self {
+            Self::Near(acc) => acc.to_string(),
+            Self::Eth {
+                receiver_id,
+                message: _,
+            } => receiver_id.to_string(),
+        }
+    }
+
+    /// Prepare message for `ft_transfer_call` -> `ft_on_transfer`
+    fn prepare_message_for_on_transfer(
+        relayer_account_id: &AccountId,
+        fee: Fee,
+        message: String,
+    ) -> Result<String, ParseEventMessageError> {
+        // First data section should contain fee data
+        let mut data = fee.into_u128().to_be_bytes().to_vec();
+
+        // Check message length.Ω
+        let address = if message.len() == 42 {
+            message
+                .strip_prefix("0x")
+                .ok_or(ParseEventMessageError::EthAddressValidationError(
+                    AddressValidationError::FailedDecodeHex,
+                ))?
+                .to_string()
+        } else {
+            message
+        };
+        let address_bytes = validate_eth_address(address)
+            .map_err(ParseEventMessageError::EthAddressValidationError)?;
+        // Second data section should contain Eth address
+        data.extend(address_bytes);
+        // Add `:` separator between relayer_id and data message
+        Ok([relayer_account_id.as_ref(), &hex::encode(data)].join(":"))
```

Comment:
I think it would be nice if there was a type like

```
struct FtOnTransferMessage {
    fee: Fee,
    address: Address,
    relayer_account_id: AccountId,
}
```

This type would be constructed in this function and passed around through the core logic. Only at the end where we are about to do the NEAR call that requires it to be a string, then we would serialize it in this way.

Reply:
Actually, we don't need that. It's used **ONLY** for:
```
pub struct TransferCallCallArgs {
    pub receiver_id: AccountId,
    pub amount: Balance,
    pub memo: Option<String>,
    pub msg: String,
}
```

Mentioned logic for `prepare_message` solves the problem with early as possible validation data. We early prepare message string for TransferCall args. I can't see any reason to introduce a new type, because we never used it.

Reply:
The reason to return structured data as a structured type instead of something opaque like a `String` is because it makes the logic more clear. Without reading the function implementation, you would never know that the `String` which is being returned is actually a structured thing.

The principle of pushing IO all the way to boundary of applications applies both to the Input and the Output sides of the IO. Yes, we want to validate input early and only pass strongly typed structures in our core logic. And similarly we want to hold structed data in strong types until the very end, where we may need to serialize it in some way to communicate with the outside world.

Reply:
With this approach, we ensure that the data is prepared for sending in the correct way. There is no problem to add an additional type, and to it a method that would generate this binary data. However, it will play the role of code readability. No more. In backward compatibility conditions, the input and output messages do not change. The new type itself will have no function other than to make the code easier to read.

Reply:
Yes, exactly. And easier to read code is a good thing.

Reply:
For that was reused and reorganized type `OnTransferMessageData`

Reply:
If `OnTransferMessageData` is now used for multiple purposes perhaps a more generic name is needed? Ie that struct is not always used with `ft_on_transfer` anymore, so its name is now misleading.